pandas/core/arrays/period.py

```diff
@@ -85,94 +108,113 @@ def wrapper(self, other):
+    """
+    Pandas ExtensionArray for storing Period data.
 
-    @cache_readonly
-    def dtype(self):
-        return PeriodDtype.construct_from_string(self.freq)
+    Users should use :func:`period_array` to create new instances.
 
-    @property
-    def _ndarray_values(self):
-        # Ordinals
-        return self._data
+    Notes
+    -----
+    There are two components to a PeriodArray
 
-    @property
-    def asi8(self):
-        return self._ndarray_values.view('i8')
+    - ordinals : integer ndarray
+    - freq : pd.tseries.offsets.Tick
 
-    @property
-    def freq(self):
-        """Return the frequency object if it is set, otherwise None"""
-        return self._freq
-
-    @freq.setter
-    def freq(self, value):
-        msg = ('Setting {cls}.freq has been deprecated and will be '
-               'removed in a future version; use {cls}.asfreq instead. '
-               'The {cls}.freq setter is not guaranteed to work.')
-        warnings.warn(msg.format(cls=type(self).__name__),
-                      FutureWarning, stacklevel=2)
-        self._freq = value
+    The values are physically stored as a 1-D ndarray of integers. These are
+    called "ordinals" and represent some kind of offset from a base.
 
-    # --------------------------------------------------------------------
-    # Constructors
+    The `freq` indicates the span covered by each element of the array.
+    All elements in the PeriodArray have the same `freq`.
 
+    See Also
+    --------
+    period_array : Create a new PeriodArray
+    pandas.PeriodIndex : Immutable Index for period data
+    """
     _attributes = ["freq"]
+    _typ = "periodarray"  # ABCPeriodArray
+
+    # Names others delegate to us
+    _other_ops = []
+    _bool_ops = ['is_leap_year']
+    _object_ops = ['start_time', 'end_time', 'freq']
+    _field_ops = ['year', 'month', 'day', 'hour', 'minute', 'second',
+                  'weekofyear', 'weekday', 'week', 'dayofweek',
+                  'dayofyear', 'quarter', 'qyear',
+                  'days_in_month', 'daysinmonth']
+    _datetimelike_ops = _field_ops + _object_ops + _bool_ops
+    _datetimelike_methods = ['strftime', 'to_timestamp', 'asfreq']
 
-    def __new__(cls, values, freq=None, **kwargs):
-        if is_period_dtype(values):
-            # PeriodArray, PeriodIndex
-            if freq is not None and values.freq != freq:
-                raise IncompatibleFrequency(freq, values.freq)
-            freq = values.freq
-            values = values.asi8
-
-        elif is_datetime64_dtype(values):
-            # TODO: what if it has tz?
-            values = dt64arr_to_periodarr(values, freq)
-
-        return cls._simple_new(values, freq=freq, **kwargs)
+    # --------------------------------------------------------------------
+    # Constructors
+    def __init__(self, values, freq=None, copy=False):
+        # type: (Union[PeriodArray, np.ndarray], Union[str, Tick]) -> None
+        if isinstance(values, ABCSeries):
```

Comment:
shouldn't this be simpler and instead just call ``period_array``?

Reply:
See the discussion in https://github.com/pandas-dev/pandas/issues/23212. From there, I understood that you were fine with keeping `__init__` limited in functionality, and having `period_array` that can do inference.

Reply:
sorry, yes you are right, this is what we do in IntegerArray. I would maybe document this very clearly, e.g. ``__init__`` is exactly construction from the underlying primitive array and has NO inference to be clear.